File: go/pgmq/consumer.go
package pgmq

import (
	"context"
	"encoding/json"
	"errors"
	"math"
	"strconv"
	"sync"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/sns45/anyq/go/core"
)

// Consumer reads pgmq messages and implements native parking and DLQ transfer.
type Consumer struct {
	*core.BaseConsumer
	cfg Config
	db  client
}

func NewConsumer(cfg Config) (*Consumer, error) {
	cfg, err := cfg.resolve()
	if err != nil {
		return nil, err
	}
	bc := &core.BaseConsumer{}
	bc.InitBase(cfg.BaseQueueConfig)
	c := &Consumer{BaseConsumer: bc, cfg: cfg}
	c.db = client{cfg: cfg, base: &bc.BaseAdapter}
	bc.Bind(c)
	return c, nil
}

func (c *Consumer) Connect(ctx context.Context) error {
	if err := c.VerifyParkPolicy(); err != nil {
		return err
	}
	return c.db.connect(ctx)
}
func (c *Consumer) Disconnect(ctx context.Context) error { return c.db.disconnect() }
func (c *Consumer) Client() *pgxpool.Pool                { return c.db.pool() }
func (c *Consumer) SupportsNativeDelay() bool            { return true }
func (c *Consumer) HealthCheck(ctx context.Context) (core.HealthStatus, error) {
	h, err := c.db.health(ctx)
	if h.Details != nil {
		h.Details["paused"] = c.IsPaused()
	}
	return h, err
}

// Record is the native pgmq read result returned by Message.Raw.
type Record struct {
	MsgID      int64
	ReadCount  int
	EnqueuedAt time.Time
	VisibleAt  time.Time
	Body       json.RawMessage
	Headers    json.RawMessage
}

type delivery struct {
	mu      sync.Mutex
	settled bool
	owner   *Consumer
	session *session
	record  Record
}

type message struct {
	core.Message
	delivery *delivery
}

func (c *Consumer) wrap(s *session, record Record) (*message, error) {
	values := map[string]json.RawMessage{}
	if len(record.Headers) > 0 && string(record.Headers) != "null" {
		if err := json.Unmarshal(record.Headers, &values); err != nil {
			return nil, core.NewSerializationError("invalid pgmq headers", err)
		}
	}
	headers := core.MessageHeaders{}
	for k, v := range values {
		var str string
		if err := json.Unmarshal(v, &str); err == nil {
			headers[k] = []byte(str)
		} else {
			headers[k] = append([]byte(nil), v...)
		}
	}
	key := string(headers[routingKeyHeader])
	delete(headers, routingKeyHeader)
	d := &delivery{owner: c, session: s, record: record}
	id := strconv.FormatInt(record.MsgID, 10)
	m := core.NewMessage(core.MessageParams{
		ID: id, Body: record.Body, Headers: headers, Key: key, Timestamp: record.EnqueuedAt, DeliveryAttempt: record.ReadCount,
		Metadata: core.ProviderMetadata{Provider: core.DriverPgmq, Pgmq: &core.PgmqMetadata{QueueName: c.cfg.QueueName, MsgID: id, ReadCount: record.ReadCount, EnqueuedAt: record.EnqueuedAt, VisibleAt: record.VisibleAt}},
		Raw:      record,
		OnAck:    func(ctx context.Context) error { return d.dispose(ctx, "delete", 0) },
		OnNack: func(ctx context.Context, requeue bool) error {
			if requeue {
				return d.dispose(ctx, "release", 0)
			}
			return d.dispose(ctx, "archive", 0)
		},
		OnExtendDeadline: func(ctx context.Context, duration time.Duration) error {
			n, err := seconds(duration)
			if err != nil {
				return err
			}
			return d.dispose(ctx, "extend", n)
		},
	})
	return &message{Message: m, delivery: d}, nil
}

func (d *delivery) dispose(ctx context.Context, action string, seconds int32) error {
	active := d.session.acquire()
	if active {
		defer d.session.release()
	}
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.settled {
		return nil
	}
	if !active {
		return core.NewConnectionError("pgmq delivery session has disconnected", nil)
	}
	ctx, cancel := context.WithTimeout(context.WithoutCancel(ctx), d.owner.cfg.RequestTimeout)
	defer cancel()
	var err error
	switch action {
	case "delete", "archive":
		var done bool
		query := "SELECT pgmq.delete($1,$2::bigint)"
		if action == "archive" {
			query = "SELECT pgmq.archive($1,$2::bigint)"
		}
		err = d.session.pool.QueryRow(ctx, query, d.owner.cfg.QueueName, d.record.MsgID).Scan(&done)
	default:
		var id int64
		err = d.session.pool.QueryRow(ctx, "SELECT msg_id FROM pgmq.set_vt($1,$2::bigint,$3::integer)", d.owner.cfg.QueueName, d.record.MsgID, seconds).Scan(&id)
		if errors.Is(err, pgx.ErrNoRows) {
			err = nil
		}
	}
	if err != nil {
		return core.NewConsumeError("failed to settle pgmq message", err)
	}
	if action != "extend" {
		d.settled = true
	}
	return nil
}

func (c *Consumer) ownDelivery(msg core.Message) (*delivery, error) {
	m, ok := msg.(*message)
	if !ok || m.delivery.owner != c {
		return nil, core.NewConfigurationError("message does not belong to this pgmq consumer", nil)
	}
	return m.delivery, nil
}

// ParkMessage returns the same message to pgmq with a future visibility time.
func (c *Consumer) ParkMessage(ctx context.Context, msg core.Message, delayMs int) error {
	d, err := c.ownDelivery(msg)
	if err != nil {
		return err
	}
	if delayMs < 0 || int64(delayMs) > int64(math.MaxInt32)*1000 {
		return core.NewConfigurationError("park delay must fit a nonnegative Postgres integer number of seconds", nil)
	}
	n := int32((int64(delayMs) + 999) / 1000)
	return d.dispose(ctx, "park", n)
}

// DeadLetterMessage atomically sends to the DLQ and removes the original.
// Failed transfers leave the original available for redelivery after its lease.
func (c *Consumer) DeadLetterMessage(ctx context.Context, msg core.Message, reason string) error {
	d, err := c.ownDelivery(msg)
	if err != nil {
		return err
	}
	active := d.session.acquire()
	if active {
		defer d.session.release()
	}
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.settled {
		return nil
	}
	if !active {
		return core.NewConnectionError("pgmq delivery session has disconnected", nil)
	}
	ctx, cancel := context.WithTimeout(context.WithoutCancel(ctx), c.cfg.RequestTimeout)
	defer cancel()
	headers := map[string]json.RawMessage{}
	if len(d.record.Headers) > 0 && string(d.record.Headers) != "null" {
		if err := json.Unmarshal(d.record.Headers, &headers); err != nil {
			return core.NewSerializationError("invalid pgmq headers", err)
		}
	}
	for k, v := range map[string]string{"x-original-queue": c.cfg.QueueName, "x-death-time": time.Now().UTC().Format(time.RFC3339Nano), "x-delivery-attempts": strconv.Itoa(msg.DeliveryAttempt())} {
		headers[k], _ = json.Marshal(v)
	}
	if dlq := c.cfg.BaseQueueConfig.DeadLetterQueue; dlq == nil || dlq.IncludeError {
		headers["x-death-reason"], _ = json.Marshal(reason)
	} else {
		delete(headers, "x-death-reason")
	}
	packed, err := json.Marshal(headers)
	if err != nil {
		return core.NewSerializationError("failed to encode DLQ headers", err)
	}
	tx, err := d.session.pool.Begin(ctx)
	if err != nil {
		return core.NewConsumeError("failed to begin pgmq DLQ transfer", err)
	}
	defer tx.Rollback(context.WithoutCancel(ctx))
	if _, err := tx.Exec(ctx, "SELECT pgmq.send($1,$2::jsonb,$3::jsonb,0)", c.cfg.DeadLetterQueue, d.record.Body, packed); err != nil {
		return core.NewConsumeError("failed to publish pgmq dead letter", err)
	}
	var deleted bool
	if err := tx.QueryRow(ctx, "SELECT pgmq.delete($1,$2::bigint)", c.cfg.QueueName, d.record.MsgID).Scan(&deleted); err != nil {
		return core.NewConsumeError("failed to remove pgmq original", err)
	}
	if !deleted {
		// The source was already removed. Roll back the duplicate DLQ copy.
		if err := tx.Rollback(ctx); err != nil {
			return core.NewConsumeError("failed to roll back pgmq DLQ transfer", err)
		}
		d.settled = true
		return nil
	}
	if err := tx.Commit(ctx); err != nil {
		return core.NewConsumeError("failed to commit pgmq DLQ transfer", err)
	}
	d.settled = true
	return nil
}

func (c *Consumer) read(ctx context.Context, s *session, size int) ([]Record, error) {
	// Let a bounded read finish even during shutdown, then release any leases
	// it returns. Cancelling the query could lose a committed read result.
	ctx, cancel := context.WithTimeout(context.WithoutCancel(ctx), c.cfg.RequestTimeout)
	defer cancel()
	vt, _ := seconds(c.cfg.VisibilityTimeout)
	rows, err := s.pool.Query(ctx, "SELECT msg_id,read_ct,enqueued_at,vt,message,headers FROM pgmq.read($1,$2::integer,$3::integer)", c.cfg.QueueName, vt, size)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var records []Record
	for rows.Next() {
		var r Record
		if err := rows.Scan(&r.MsgID, &r.ReadCount, &r.EnqueuedAt, &r.VisibleAt, &r.Body, &r.Headers); err != nil {
			return records, err
		}
		records = append(records, r)
	}
	return records, rows.Err()
}

func (c *Consumer) release(ctx context.Context, s *session, records []Record) error {
	ctx, cancel := context.WithTimeout(context.WithoutCancel(ctx), c.cfg.RequestTimeout)
	defer cancel()
	var result error
	for _, r := range records {
		_, err := s.pool.Exec(ctx, "SELECT pgmq.set_vt($1,$2::bigint,0)", c.cfg.QueueName, r.MsgID)
		if err != nil {
			result = errors.Join(result, core.NewConsumeError("failed to release undelivered pgmq lease", err))
		}
	}
	return result
}

func sleep(ctx context.Context, d time.Duration) error {
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

func (c *Consumer) loop(ctx context.Context, s *session, size int, handle func([]core.Message) error) error {
	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		if c.IsPaused() {
			if err := sleep(ctx, c.cfg.PollInterval); err != nil {
				return err
			}
			continue
		}
		records, err := c.read(ctx, s, size)
		if err != nil {
			if releaseErr := c.release(ctx, s, records); releaseErr != nil {
				return errors.Join(err, releaseErr)
			}
			if ctx.Err() != nil {
				return ctx.Err()
			}
			ce := core.NewConsumeError("failed to read pgmq messages", err)
			c.Logger.Error(ce.Error(), nil)
			if c.Config.OnError != nil {
				c.Config.OnError(ce)
			}
			if err := sleep(ctx, c.cfg.PollInterval); err != nil {
				return err
			}
			continue
		}
		if ctx.Err() != nil || c.IsPaused() {
			if err := c.release(ctx, s, records); err != nil {
				return err
			}
			continue
		}
		if len(records) == 0 {
			if err := sleep(ctx, c.cfg.PollInterval); err != nil {
				return err
			}
			continue
		}
		msgs := make([]core.Message, 0, len(records))
		for _, r := range records {
			m, err := c.wrap(s, r)
			if err != nil {
				return errors.Join(err, c.release(ctx, s, records))
			}
			msgs = append(msgs, m)
		}
		if err := handle(msgs); err != nil {
			return err
		}
	}
}

// Subscribe runs at most Concurrency handlers at once, each with its own read.
func (c *Consumer) Subscribe(ctx context.Context, handler core.Handler, opts *core.SubscribeOptions) error {
	if handler == nil {
		return core.NewConfigurationError("handler is required", nil)
	}
	s, err := c.db.acquire()
	if err != nil {
		return err
	}
	defer s.release()
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
	stop := context.AfterFunc(s.ctx, cancel)
	defer stop()
	workers := 1
	if opts != nil && opts.Concurrency > 0 {
		workers = opts.Concurrency
	}
	results := make(chan error, workers)
	for i := 0; i < workers; i++ {
		go func() {
			results <- c.loop(ctx, s, 1, func(msgs []core.Message) error {
				m := msgs[0]
				err := handler(ctx, m)
				if err == nil {
					if opts.AutoAckEnabled() {
						return m.Ack(ctx)
					}
					return nil
				}
				return c.failed(ctx, m, err, func() error { return handler(ctx, m) })
			})
		}()
	}
	var result error
	for i := 0; i < workers; i++ {
		err := <-results
		if result == nil || (errors.Is(result, context.Canceled) && !errors.Is(err, context.Canceled)) {
			result = err
		}
		cancel()
	}
	return result
}

func (c *Consumer) SubscribeBatch(ctx context.Context, handler core.BatchHandler, opts *core.SubscribeOptions) error {
	if handler == nil {
		return core.NewConfigurationError("batch handler is required", nil)
	}
	size := c.cfg.BatchSize
	if opts != nil && opts.BatchSize > 0 {
		size = opts.BatchSize
	}
	if size > math.MaxInt32 {
		return core.NewConfigurationError("batch size must fit a Postgres integer", nil)
	}
	s, err := c.db.acquire()
	if err != nil {
		return err
	}
	defer s.release()
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
	stop := context.AfterFunc(s.ctx, cancel)
	defer stop()
	return c.loop(ctx, s, size, func(msgs []core.Message) error {
		herr := handler(ctx, msgs)
		for _, m := range msgs {
			if herr == nil {
				if opts.AutoAckEnabled() {
					if err := m.Ack(ctx); err != nil {
						return err
					}
				}
			} else {
				if err := c.failed(ctx, m, herr, func() error { return handler(ctx, []core.Message{m}) }); err != nil {
					return err
				}
			}
		}
		return nil
	})
}

func (c *Consumer) failed(ctx context.Context, msg core.Message, herr error, reinvoke func() error) error {
	d, _ := c.ownDelivery(msg)
	d.mu.Lock()
	settled := d.settled
	d.mu.Unlock()
	if settled {
		return nil
	}
	handled, err := c.ApplyStrategy(ctx, msg, herr, reinvoke)
	if handled || err != nil {
		return err
	}
	c.Logger.Error("error processing pgmq message", map[string]any{"messageId": msg.ID(), "error": herr.Error()})
	if c.Config.OnError != nil {
		c.Config.OnError(herr)
	}
	if dlq := c.cfg.BaseQueueConfig.DeadLetterQueue; dlq != nil && dlq.Enabled && msg.DeliveryAttempt() >= c.ResolveMaxAttempts() {
		return c.DeadLetterMessage(ctx, msg, herr.Error())
	}
	return nil
}

var _ core.Consumer = (*Consumer)(nil)
var _ core.ConsumerHooks = (*Consumer)(nil)

File: packages/pgmq/src/consumer.ts
/**
 * @fileoverview pgmq consumer implementation
 * @module @anyq/pgmq/consumer
 */

import type { Pool } from 'pg';
import {
  BaseConsumer,
  createMessage,
  type MessageHandler,
  type BatchMessageHandler,
  type SubscribeOptions,
  type HealthStatus,
  type IMessage,
  ConfigurationError,
  ConnectionError,
} from '@anyq/core';
import type { PgmqConfig, PgmqConsumerOptions } from './config.js';
import {
  asConnectionError,
  createPool,
  ensurePgmq,
  ensureQueue,
  hasReadWithPoll,
  packHeaders,
  toJsonText,
  unpackHeaders,
  validateQueueName,
  type PgmqRow,
} from './client.js';

/**
 * Default subscribe options
 */
const DEFAULT_SUBSCRIBE_OPTIONS: Required<SubscribeOptions> = {
  fromBeginning: false,
  fromTimestamp: undefined as unknown as Date,
  concurrency: 1,
  autoAck: true,
  batchSize: 10,
  batchTimeout: 1000,
};

/**
 * pgmq consumer implementation
 *
 * Reads with `pgmq.read` (or `pgmq.read_with_poll` when available) using the
 * visibility timeout as the processing lease, exactly like SQS. A message
 * that is neither acked nor nacked reappears once its `vt` lapses, with
 * `read_ct` incremented, so `deliveryAttempt` is always truthful.
 *
 * Native hooks:
 * - `park`: `pgmq.set_vt(queue, id, delay)`, no in-process downgrade
 * - dead letter: `pgmq.send` to the DLQ queue, then delete the original
 * - `nack(false)`: `pgmq.archive`, so nothing is silently dropped
 *
 * @example
 * ```typescript
 * const consumer = new PgmqConsumer({
 *   driver: 'pgmq',
 *   queueName: 'orders',
 *   pg: { connectionString: 'postgres://postgres:postgres@localhost:5432/postgres' },
 *   consumer: { visibilityTimeout: 30 },
 * });
 *
 * await consumer.connect();
 * await consumer.subscribe(async (message) => {
 *   console.log('Received:', message.body);
 *   await message.ack();
 * });
 * ```
 */
export class PgmqConsumer<T = unknown> extends BaseConsumer<T> {
  private pool: Pool | null = null;
  private ownsPool = false;
  private readonly queueName: string;
  private readonly dlqName: string | null;
  private readonly consumerOptions: Required<PgmqConsumerOptions>;
  private running = false;
  private pollingPromise: Promise<void> | null = null;
  private useLongPoll = false;

  // pgmq schedules redelivery natively through the visibility timeout, so
  // `park` never downgrades to an in-process sleep.
  protected get supportsNativeDelay(): boolean {
    return true;
  }

  constructor(config: PgmqConfig) {
    super(config);
    this.queueName = validateQueueName(config.queueName);

    const dlq = config.deadLetterQueue;
    this.dlqName = dlq?.enabled
      ? validateQueueName(dlq.destination || `${this.queueName}_dlq`, 'deadLetterQueue.destination')
      : null;

    const c = config.consumer ?? {};
    this.consumerOptions = {
      visibilityTimeout: c.visibilityTimeout ?? 30,
      pollingInterval: c.pollingInterval ?? 1000,
      longPollSeconds: c.longPollSeconds ?? 5,
      longPollIntervalMs: c.longPollIntervalMs ?? 100,
      maxMessages: c.maxMessages ?? 100,
    };
  }

  /**
   * Open the pool, verify pgmq, create the queue and DLQ when allowed, and
   * detect long poll support.
   */
  async connect(): Promise<void> {
    if (this._connected) {
      return;
    }

    const config = this.config as PgmqConfig;
    const { pool, owned } = createPool(config);
    pool.on('error', (err) => {
      this.logger.error('pg pool error', { error: err.message });
    });

    try {
      await ensurePgmq(pool, config.autoInstall ?? true);
      if (config.autoCreate !== false) {
        await ensureQueue(pool, this.queueName);
        if (this.dlqName) {
          await ensureQueue(pool, this.dlqName);
        }
      }
      this.useLongPoll =
        this.consumerOptions.longPollSeconds > 0 && (await hasReadWithPoll(pool));
      if (this.consumerOptions.longPollSeconds > 0 && !this.useLongPoll) {
        this.logger.debug(
          'pgmq.read_with_poll not available; using pgmq.read with pollingInterval',
        );
      }
    } catch (error) {
      if (owned) {
        await pool.end().catch(() => undefined);
      }
      if (error instanceof ConnectionError || error instanceof ConfigurationError) {
        throw error;
      }
      throw asConnectionError('Failed to connect to Postgres', error);
    }

    this.pool = pool;
    this.ownsPool = owned;
    this._connected = true;
    this.logger.info('pgmq consumer connected', {
      queue: this.queueName,
      dlq: this.dlqName ?? undefined,
      visibilityTimeout: this.consumerOptions.visibilityTimeout,
      longPoll: this.useLongPoll,
    });

    // Native delay, so this only re-arms the shutdown flag; kept uniform
    // across adapters.
    this.verifyParkPolicy();
  }

  /**
   * Stop polling and end the pool (only if this adapter created it).
   */
  async disconnect(): Promise<void> {
    this.beginShutdown();
    if (!this._connected || !this.pool) {
      return;
    }

    this.running = false;
    if (this.pollingPromise) {
      await this.pollingPromise;
      this.pollingPromise = null;
    }

    const pool = this.pool;
    this.pool = null;
    this._connected = false;
    if (this.ownsPool) {
      await pool.end().catch(() => undefined);
    }
    this.logger.info('pgmq consumer disconnected');
  }

  /**
   * Subscribe to messages
   */
  async subscribe(
    handler: MessageHandler<T>,
    options?: SubscribeOptions,
  ): Promise<void> {
    if (!this.pool) {
      throw new ConnectionError('Consumer not connected');
    }

    const opts = { ...DEFAULT_SUBSCRIBE_OPTIONS, ...options };
    this.running = true;
    this.pollingPromise = this.poll(handler, opts);

    this.logger.info('Subscribed to queue', { queue: this.queueName });
  }

  /**
   * Subscribe to message batches
   */
  async subscribeBatch(
    handler: BatchMessageHandler<T>,
    options?: SubscribeOptions,
  ): Promise<void> {
    if (!this.pool) {
      throw new ConnectionError('Consumer not connected');
    }

    const opts = { ...DEFAULT_SUBSCRIBE_OPTIONS, ...options };
    this.running = true;
    this.pollingPromise = this.pollBatch(handler, opts);

    this.logger.info('Subscribed to queue (batch)', { queue: this.queueName });
  }

  /**
   * Native park: push the visibility timeout out by `delayMs`. The message
   * comes back with `read_ct` incremented, so the next attempt is counted.
   */
  protected override async parkMessage(
    message: IMessage<T>,
    delayMs: number,
  ): Promise<void> {
    const seconds = Math.max(0, Math.ceil(delayMs / 1000));
    try {
      await this.setVt(message.id, seconds);
      this.logger.debug('Message parked for redelivery', {
        messageId: message.id,
        delaySeconds: seconds,
      });
    } catch (err) {
      this.logger.error('Failed to park message; returning it to the queue', {
        messageId: message.id,
        error: err instanceof Error ? err.message : String(err),
      });
      await message.nack(true);
    }
  }

  /**
   * Native dead letter: copy to the DLQ queue with the same headers the
   * memory adapter attaches, then delete the original. Without a DLQ the
   * message is archived rather than dropped.
   */
  protected override async deadLetterMessage(
    message: IMessage<T>,
    reason: string,
  ): Promise<void> {
    const pool = this.pool;
    if (this.dlqName && pool) {
      const dlqConfig = this.config.deadLetterQueue;
      const headers: Record<string, string | Buffer | undefined> = {
        ...message.headers,
        'x-original-queue': this.queueName,
        'x-death-time': new Date().toISOString(),
        'x-delivery-attempts': String(message.deliveryAttempt),
      };
      if (dlqConfig?.includeError !== false) {
        headers['x-death-reason'] = reason;
      }

      try {
        await pool.query('SELECT pgmq.send($1, $2::jsonb, $3::jsonb)', [
          this.dlqName,
          toJsonText(this.serializer.serialize(message.body)),
          JSON.stringify(packHeaders(headers, message.key)),
        ]);
        await this.deleteOne(message.id);
        this.logger.warn('Message dead-lettered', {
          messageId: message.id,
          reason,
          dlq: this.dlqName,
        });
        return;
      } catch (err) {
        this.logger.error('Failed to dead-letter message; archiving instead', {
          messageId: message.id,
          error: err instanceof Error ? err.message : String(err),
        });
      }
    } else {
      this.logger.warn('No DLQ configured; archiving message', {
        messageId: message.id,
        reason,
      });
    }
    await message.nack(false);
  }

  /**
   * Round trip to Postgres plus queue depth from `pgmq.metrics`.
   */
  async healthCheck(): Promise<HealthStatus> {
    const start = Date.now();
    if (!this.pool || !this._connected) {
      return { healthy: false, connected: false, error: 'Not connected' };
    }
    try {
      await this.pool.query('SELECT 1');
      const details: Record<string, unknown> = {
        queue: this.queueName,
        dlq: this.dlqName ?? undefined,
        paused: this._paused,
        running: this.running,
        longPoll: this.useLongPoll,
      };
      try {
        const { rows } = await this.pool.query<{ queue_length: string; total_messages: string }>(
          'SELECT queue_length, total_messages FROM pgmq.metrics($1)',
          [this.queueName],
        );
        if (rows[0]) {
          details.queueLength = Number(rows[0].queue_length);
          details.totalMessages = Number(rows[0].total_messages);
        }
      } catch {
        // metrics are best effort
      }
      return { healthy: true, connected: true, latencyMs: Date.now() - start, details };
    } catch (error) {
      return {
        healthy: false,
        connected: false,
        latencyMs: Date.now() - start,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  /**
   * Get the underlying pool (for testing)
   */
  getPool(): Pool | null {
    return this.pool;
  }

  /**
   * Single message poll loop
   */
  private async poll(
    handler: MessageHandler<T>,
    opts: Required<SubscribeOptions>,
  ): Promise<void> {
    const qty = Math.min(Math.max(1, opts.concurrency), this.consumerOptions.maxMessages);

    while (this.running) {
      if (this._paused) {
        await this.sleep(this.consumerOptions.pollingInterval);
        continue;
      }

      try {
        const rows = await this.read(qty);
        if (rows.length === 0) {
          if (!this.useLongPoll) {
            await this.sleep(this.consumerOptions.pollingInterval);
          }
          continue;
        }

        // A long poll may return after pause() or disconnect() was called;
        // hand those leases straight back instead of invoking the handler.
        if (this._paused || !this.running) {
          await this.release(rows);
          continue;
        }

        for (let i = 0; i < rows.length; i++) {
          if (!this.running || this._paused) {
            await this.release(rows.slice(i));
            break;
          }
          await this.handleOne(rows[i], handler, opts);
        }
      } catch (error) {
        if (this.running) {
          this.logger.error('Error polling messages', {
            error: error instanceof Error ? error.message : String(error),
          });
          await this.sleep(this.consumerOptions.pollingInterval);
        }
      }
    }
  }

  /**
   * Batch poll loop
   */
  private async pollBatch(
    handler: BatchMessageHandler<T>,
    opts: Required<SubscribeOptions>,
  ): Promise<void> {
    const qty = Math.min(Math.max(1, opts.batchSize), this.consumerOptions.maxMessages);

    while (this.running) {
      if (this._paused) {
        await this.sleep(this.consumerOptions.pollingInterval);
        continue;
      }

      try {
        const rows = await this.read(qty);
        if (rows.length === 0) {
          if (!this.useLongPoll) {
            await this.sleep(this.consumerOptions.pollingInterval);
          }
          continue;
        }

        if (this._paused || !this.running) {
          await this.release(rows);
          continue;
        }

        const messages = rows.map((row) => this.createWrappedMessage(row));

        try {
          await handler(messages);
          if (opts.autoAck) {
            await this.deleteMany(messages.map((m) => m.id));
          }
        } catch (error) {
          const err = error instanceof Error ? error : new Error(String(error));
          // pgmq acks per message, so apply the strategy per message.
          for (const message of messages) {
            const result = await this.applyStrategy(message, err, () => handler([message]));
            if (!result.handled) {
              await this.legacyFailure(message, err);
            }
          }
        }
      } catch (error) {
        if (this.running) {
          this.logger.error('Error polling batch', {
            error: error instanceof Error ? error.message : String(error),
          });
          await this.sleep(this.consumerOptions.pollingInterval);
        }
      }
    }
  }

  /**
   * Deliver one row to the handler and route any failure through the
   * configured strategy.
   */
  private async handleOne(
    row: PgmqRow,
    handler: MessageHandler<T>,
    opts: Required<SubscribeOptions>,
  ): Promise<void> {
    const message = this.createWrappedMessage(row);
    try {
      this.emit('message', message);
      await handler(message);
      if (opts.autoAck) {
        await message.ack();
      }
    } catch (error) {
      const err = error instanceof Error ? error : new Error(String(error));
      const result = await this.applyStrategy(message, err, () => handler(message));
      if (!result.handled) {
        await this.legacyFailure(message, err);
      }
    }
  }

  /**
   * Behaviour when no `strategy` is configured: dead letter once
   * `deadLetterQueue.maxDeliveryAttempts` is reached, otherwise leave the
   * message alone so it reappears when its visibility timeout lapses.
   */
  private async legacyFailure(message: IMessage<T>, err: Error): Promise<void> {
    this.logger.error('Error processing message', {
      messageId: message.id,
      deliveryAttempt: message.deliveryAttempt,
      error: err.message,
    });
    this.emit('error', err);

    const max = this.config.deadLetterQueue?.maxDeliveryAttempts;
    if (this.dlqName && typeof max === 'number' && max > 0 && message.deliveryAttempt >= max) {
      await this.deadLetterMessage(message, err.message);
    }
  }

  /**
   * Read up to `qty` messages, leasing them for the visibility timeout.
   */
  private async read(qty: number): Promise<PgmqRow[]> {
    const pool = this.pool;
    if (!pool) {
      return [];
    }
    const vt = this.consumerOptions.visibilityTimeout;

    if (this.useLongPoll) {
      const { rows } = await pool.query<PgmqRow>(
        'SELECT * FROM pgmq.read_with_poll($1, $2::int, $3::int, $4::int, $5::int)',
        [
          this.queueName,
          vt,
          qty,
          this.consumerOptions.longPollSeconds,
          this.consumerOptions.longPollIntervalMs,
        ],
      );
      return rows;
    }

    const { rows } = await pool.query<PgmqRow>(
      'SELECT * FROM pgmq.read($1, $2::int, $3::int)',
      [this.queueName, vt, qty],
    );
    return rows;
  }

  /**
   * Return leased rows to the queue immediately (visibility timeout 0), used
   * when a read completes after pause() or disconnect().
   */
  private async release(rows: PgmqRow[]): Promise<void> {
    for (const row of rows) {
      try {
        await this.setVt(String(row.msg_id), 0);
      } catch (err) {
        this.logger.warn('Failed to release leased message; it will reappear after vt', {
          messageId: String(row.msg_id),
          error: err instanceof Error ? err.message : String(err),
        });
      }
    }
  }

  private async deleteOne(msgId: string): Promise<void> {
    await this.pool?.query('SELECT pgmq.delete($1, $2::bigint)', [this.queueName, msgId]);
  }

  private async deleteMany(msgIds: string[]): Promise<void> {
    if (msgIds.length === 0) return;
    await this.pool?.query('SELECT pgmq.delete($1, $2::bigint[])', [this.queueName, msgIds]);
  }

  private async setVt(msgId: string, seconds: number): Promise<void> {
    await this.pool?.query('SELECT pgmq.set_vt($1, $2::bigint, $3::int)', [
      this.queueName,
      msgId,
      Math.max(0, Math.floor(seconds)),
    ]);
  }

  private async archiveOne(msgId: string): Promise<void> {
    await this.pool?.query('SELECT pgmq.archive($1, $2::bigint)', [this.queueName, msgId]);
  }

  /**
   * Wrap a pgmq row in the universal message interface.
   */
  private createWrappedMessage(row: PgmqRow): IMessage<T> {
    const msgId = String(row.msg_id);
    // pg parses jsonb into a value; re-stringify so the serializer contract
    // (string in, T out) holds for custom serializers too.
    const body = this.serializer.deserialize(JSON.stringify(row.message ?? null));
    const { headers, key } = unpackHeaders(row.headers);
    const readCount = Number(row.read_ct) || 1;
    const enqueuedAt = new Date(row.enqueued_at);
    const visibleAt = new Date(row.vt);

    return createMessage<T>({
      id: msgId,
      body,
      key,
      headers,
      timestamp: enqueuedAt,
      deliveryAttempt: readCount,
      metadata: {
        provider: 'pgmq',
        pgmq: {
          queueName: this.queueName,
          msgId,
          readCount,
          enqueuedAt,
          visibleAt,
        },
      },
      raw: row,
      onAck: async () => {
        await this.deleteOne(msgId);
        this.logger.debug('Message acknowledged', { messageId: msgId });
      },
      onNack: async (requeue = true) => {
        if (requeue) {
          await this.setVt(msgId, 0);
        } else {
          await this.archiveOne(msgId);
        }
        this.logger.debug('Message nacked', { messageId: msgId, requeue });
      },
      onExtendDeadline: async (seconds: number) => {
        // pgmq sets an absolute deadline from now, like SQS ChangeMessageVisibility.
        await this.setVt(msgId, seconds);
        this.logger.debug('Deadline extended', { messageId: msgId, seconds });
      },
    });
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
